modules/highgui/src/window_wayland.cpp

```diff
@@ -0,0 +1,2482 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+
+#ifndef _WIN32
+#if defined (HAVE_WAYLAND)
+
+#include <iostream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <map>
+#include <queue>
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <chrono>
+#include <unordered_map>
+
+#include <unistd.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <sys/epoll.h>
+#include <linux/input.h>
+
+#include <xkbcommon/xkbcommon.h>
+#include <wayland-client-protocol.h>
+#include <wayland-cursor.h>
+#include <wayland-util.h>
+#include "xdg-shell-client-protocol.h"
+
+#include <opencv2/core.hpp>
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc.hpp>
+#include "opencv2/core/utils/logger.hpp"
+
+/*                              */
+/*  OpenCV highgui internals    */
+/*                              */
+class cv_wl_display;
+
+class cv_wl_mouse;
+
+class cv_wl_keyboard;
+
+class cv_wl_input;
+
+class cv_wl_buffer;
+
+struct cv_wl_cursor;
+
+class cv_wl_cursor_theme;
+
+class cv_wl_widget;
+
+class cv_wl_titlebar;
+
+class cv_wl_viewer;
+
+class cv_wl_trackbar;
+
+class cv_wl_window;
+
+class cv_wl_core;
+
+using std::weak_ptr;
+using std::shared_ptr;
+using namespace cv::Error;
+namespace ch = std::chrono;
+
+#define throw_system_error(errmsg, errnum) \
+    CV_Error_(StsInternal, ("%s: %s", errmsg, strerror(errnum)));
+
+// workaround for Wayland macro not compiling in C++
+#define WL_ARRAY_FOR_EACH(pos, array, type)                             \
+  for ((pos) = (type)(array)->data;                                       \
+       (const char*)(pos) < ((const char*)(array)->data + (array)->size); \
+       (pos)++)
+
+static int xkb_keysym_to_ascii(xkb_keysym_t keysym) {
+    return static_cast<int>(keysym & 0xff);
+}
+
+/*
+ * From /usr/include/wayland-client-protocol.h
+ * @WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1: libxkbcommon compatible; to
+ *  determine the xkb keycode, clients must add 8 to the key event keycode
+ */
+static xkb_keycode_t xkb_keycode_from_raw_keycode(int raw_keycode) {
+    return raw_keycode + 8;
+}
+
+static void draw_argb8888(void *d, uint8_t a, uint8_t r, uint8_t g, uint8_t b) {
+    *((uint32_t *) d) = ((a << 24) | (r << 16) | (g << 8) | b);
+}
+
+static void write_mat_to_xrgb8888(cv::Mat const &img, void *data) {
+    assert(data != nullptr);
+    assert(img.isContinuous());
```

Comment:
Please avoid using of C-like `assert()`.

Please use:
- CV_Assert() - works in both Debug/Release modes
- or CV_DbgAssert() - Debug only